lib/index.ts

```diff
@@ -0,0 +1,244 @@
+const _SIGINT = 'SIGINT';
+const _SIGINT_EXIT_CODE = 130;
+
+class PWMetrics {
+  flags: FeatureFlags|any;
+  runs: number;
+  sheets: Object;
+  expectations: Object;
+  clientSecret: AuthorizeCredentials;
+  tryLighthouseCounter: number;
+  launcher: any;
+
+  constructor(public url: string, opts: MainOptions) {
+    this.flags = opts.flags || {};
+    // @todo remove when new lighthouse version be released, because -https://github.com/GoogleChrome/lighthouse/pull/1778
+    this.flags.disableCpuThrottling = this.flags.disableCpuThrottling || false;
+    this.runs = this.flags.runs || 1;
+    this.sheets = opts.sheets;
+    this.expectations = opts.expectations || false;
+    this.clientSecret = opts.clientSecret;
+    this.tryLighthouseCounter = 0;
+
+    if (this.flags.expectations) {
+      if (this.expectations) {
+        expectations.validateMetrics(this.expectations);
+        this.expectations = expectations.normalizeMetrics(this.expectations);
+      } else throw new Error(getMessageWithPrefix('ERROR', 'NO_EXPECTATIONS_FOUND'));
+    }
+  }
+
+  async start() {
+    const runs = Array.apply(null, {length: this.runs}).map(Number.call, Number);
+    let results: MetricsResults[] = [];
+
+    // Kill spawned Chrome process in case of ctrl-C.
+    process.on(_SIGINT, async () => {
+      if (this.launcher) {
+        await this.launcher.kill();
+        process.exit(_SIGINT_EXIT_CODE);
+        console.log(messages.getMessage('CLOSING_CHROME'));
+      }
+    });
+
+    for (let runIndex of runs) {
+      try {
+        results[runIndex] = await this.run();
+        console.log(messages.getMessageWithPrefix('SUCCESS', 'SUCCESS_RUN', runIndex, runs.length));
+      } catch(error) {
+        results[runIndex] = error;
+        console.error(messages.getMessageWithPrefix('ERROR', 'FAILED_RUN', runIndex, runs.length, error.message));
+      }
+    }
+
+    const ret = {runs: results.filter(r => !(r instanceof Error))};
+    if (this.runs > 1 && !this.flags.submit) {
+      const median = this.findMedianRun(results);
+      console.log(messages.getMessage('MEDIAN_RUN'));
+      this.displayOutput(median);
+    } else if (this.flags.submit) {
+      const sheets = new Sheets(this.sheets, this.clientSecret);
+      return sheets.appendResults(results).then(() => results);
+    }
+    return ret;
+  }
+
+  async run(): Promise<MetricsResults> {
+    try {
+      let lhResults: LighthouseResults;
+      await this.launchChrome();
+
+      if (process.env.CI) {
+        // handling CRI_TIMEOUT issue - https://github.com/GoogleChrome/lighthouse/issues/833
+        this.tryLighthouseCounter = 0;
+        lhResults = await this.runLighthouseOnCI();
+      } else {
+        lhResults = await lighthouse(this.url, this.flags);
+      }
+
+      const metricsResults: MetricsResults = await this.recordLighthouseTrace(lhResults);
+      await this.launcher.kill();
+
+      return metricsResults;
+    } catch (error) {
+      throw new Error(error);
```

Comment:
Should we not kill the google chrome instance (this.launcher) if there is an error? If we don't, won't there be a zombie google chrome process? In the past I've had google chrome not closed because of a failed run.